### DISCUSSION TITLE
when i use localstorage  the show the error

### DISCUSSION THREAD

**Discussion by BADHON512** [vercel/next.js#48042] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
layout component i use "use client"
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

 when i use localstorage  the show the error

### To Reproduce

 when i use localstorage  the show the error

### Describe the Bug

 when i use localstorage  the show the error

### Expected Behavior

error - redux\Reducer\reducer.js (6:8) @ localStorage
error - ReferenceError: localStorage is not defined
    at eval (./redux/Reducer/reducer.js:7:11)
    at (sc_client)/./redux/Reducer/reducer.js (E:\NEXT.JS\xman\.next\server\app\cart\page.js:1957:1)
    at __webpack_require__ (E:\NEXT.JS\xman\.next\server\webpack-runtime.js:33:43)
    at eval (./redux/Store/store.js:5:74)
    at (sc_client)/./redux/Store/store.js (E:\NEXT.JS\xman\.next\server\app\cart\page.js:1968:1)
    at __webpack_require__ (E:\NEXT.JS\xman\.next\server\webpack-runtime.js:33:43)
    at eval (./GlobalProvider/GlobalProvider.js:7:76)
    at (sc_client)/./GlobalProvider/GlobalProvider.js (E:\NEXT.JS\xman\.next\server\app\cart\page.js:736:1)
    at __webpack_require__ (E:\NEXT.JS\xman\.next\server\webpack-runtime.js:33:43)
  4 | 
  5 | const initialState = {
> 6 |   Cart: localStorage.getItem("Cart")?JSON.parse(localStorage.getItem("Cart")):{CartItem:[]}
    |        ^
  7 | };
  8 |
  9 |

**Comment by OgDev-01** [comment]
@BADHON512 can you provide a bit more context on the current issue? e.g: a link to the code that produced the issue or a screenshot to help track the problem. 🍕

**Comment by BADHON512** [comment]
![image](https://user-images.githubusercontent.com/93506264/230002777-4dadb24d-97eb-4e27-a3d3-c58252348726.png)

**Comment by OgDev-01** [comment]
Did you check to see if the client window is not equals to undefined before calling localstorage?


### Reason:
The `localStorage` object is only available in certain environments, such as web browsers. If you try to access the localStorage object in an environment where it is not available, you will get an error.

To prevent this error from occurring, it is common practice to check if the localStorage object is available before using it. One way to do this is to check if the typeof the window object is not equal to undefined, since the window object is only available in a browser environment.

#### Here's an example of how this check can be performed:

```javascript
if (typeof window !== 'undefined') {
  // window object is available, so localStorage should be too
  // your localstorage code should go here
} else {
  // window object is not available, so localStorage cannot be used
  console.log('localStorage is not available');
}
```

I don't think this issue is related to Nextjs. 🍕

**Comment by SukkaW** [comment]
@OgDev-01 explains the issue well. And I'd like to share more details here.

The client components (that you have marked with the "use client" directives) don't mean they will only be evaluated on the client. They will still be evaluated on the server for SSR. That's why you still want to guard your DOM/Web API access with `typeof window`.

**Comment by brunocroh** [comment]
@SukkaW this info exists on next.js docs?

I read the docs about client component and no find details about that

Tks for clarifying!

**Comment by samcx** [comment]
Moving this issue to discussions as this is a code-related issue, as pointed out by @SukkaW!
